File: 0059_spiral_matrix_ii.swift
// https://leetcode.com/problems/spiral-matrix-ii/

class Solution {
	private enum Direction {
		case left, right, up, down
	}

	func generateMatrix(_ n: Int) -> [[Int]] {
		let row = Array(repeating: 0, count: n)
		var matrix = Array(repeating: row, count: n)

		var i = 0
		var j = -1
		var direction = Direction.right

		for num in 1...(n * n) {
			switch direction {
				case .right:
					if j + 1 == n || matrix[i][j + 1] != 0 {
						direction = .down
						i += 1
					} else {
						j += 1
					}
				case .down:
					if i + 1 == n || matrix[i + 1][j] != 0 {
						direction = .left
						j -= 1
					} else {
						i += 1
					}
				case .left:
					if j - 1 < 0 || matrix[i][j - 1] != 0 {
						direction = .up
						i -= 1
					} else {
						j -= 1
					}
				case .up:
					if i - 1 < 0 || matrix[i - 1][j] != 0 {
						direction = .right
						j += 1
					} else {
						i -= 1
					}
			}

			matrix[i][j] = num
		}

		return matrix
	}
}

File: 0059_spiral_matrix_ii.kt
// https://leetcode.com/problems/spiral-matrix-ii/description/

class Solution {
    fun generateMatrix(n: Int): Array<IntArray> {
        val matrix = Array(n) { IntArray(n) }
        var y = 0
        var x = 0
        var direction = Direction.RIGHT
        for (i in 1..n * n) {
            matrix[y][x] = i
            when (direction) {
                Direction.RIGHT -> {
                    if (x + 1 >= n || matrix[y][x + 1] != 0) {
                        direction = Direction.DOWN
                        y++
                    } else {
                        x++
                    }
                }
                Direction.DOWN -> {
                    if (y + 1 >=n || matrix[y + 1][x] != 0) {
                        direction = Direction.LEFT
                        x--
                    } else {
                        y++
                    }
                }
                Direction.LEFT -> {
                    if (x - 1 < 0 || matrix[y][x - 1] != 0) {
                        direction = Direction.UP
                        y--
                    } else {
                        x--
                    }
                }
                Direction.UP -> {
                    if (y - 1 < 0 || matrix[y - 1][x] != 0) {
                        direction = Direction.RIGHT
                        x++
                    } else {
                        y--
                    }
                }
            }
        }
        return matrix
    }
}

enum class Direction {
    LEFT, RIGHT, UP, DOWN
}
